vanetza/security/default_certificate_validator.cpp

```diff
@@ -0,0 +1,122 @@
+    // section 6.7 in TS 103 097 v1.2.1
+    unsigned certificate_time_constraints = 0;
+
+    // check validity restriction
+    for (auto& restriction : certificate.validity_restriction) {
+        ValidityRestriction validity_restriction = restriction;
+        ValidityRestrictionType type = get_type(validity_restriction);
+
+        if (type == ValidityRestrictionType::Time_Start_And_End) {
+            // change start and end time of certificate validity
+            StartAndEndValidity start_and_end = boost::get<StartAndEndValidity>(validity_restriction);
+
+            // check if certificate validity restriction timestamps are logically correct
+            if (start_and_end.start_validity >= start_and_end.end_validity) {
+                return CertificateInvalidReason::BROKEN_TIME_PERIOD;
+            }
+
+            // check if certificate is premature or outdated
+            auto now = convert_time32(time_now);
+            if (now < start_and_end.start_validity || now > start_and_end.end_validity) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        } else if (type == ValidityRestrictionType::Time_End) {
+            EndValidity end = boost::get<EndValidity>(validity_restriction);
+
+            // check if certificate is outdated
+            auto now = convert_time32(time_now);
+            if (now > end) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        } else if (type == ValidityRestrictionType::Time_Start_And_Duration) {
+            StartAndDurationValidity start_and_duration = boost::get<StartAndDurationValidity>(validity_restriction);
+
+            // check if certificate is premature or outdated
+            auto now = convert_time32(time_now);
+            std::chrono::seconds duration = start_and_duration.duration.to_seconds();
+            auto end = start_and_duration.start_validity + duration.count();
+            if (now < start_and_duration.start_validity || now > end) {
+                return CertificateInvalidReason::OFF_TIME_PERIOD;
+            }
+
+            ++certificate_time_constraints;
+        }
+    }
+
+    // if not exactly one time constraint is given, we fail instead of considering it valid
+    if (1 != certificate_time_constraints) {
+        return CertificateInvalidReason::BROKEN_TIME_PERIOD;
+    }
+
+    // check if subject_name is empty
+    if (0 != certificate.subject_info.subject_name.size()) {
+        return CertificateInvalidReason::INVALID_NAME;
+    }
+
+    // check signer info
+    if (get_type(certificate.signer_info) != SignerInfoType::Certificate_Digest_With_SHA256) {
+        return CertificateInvalidReason::INVALID_ROOT_HASH;
+    }
+
+    HashedId8 signer_hash = boost::get<HashedId8>(certificate.signer_info);
+    std::vector<Certificate> possible_signers = trust_store.find_by_id(signer_hash);
```

Comment:
Wouldn't it be better to fetch the claimed signer from the certificate and then verify whether that signer is actually valid? If I understand this part correctly, you're taking the signers from the trust store and checking whether you have the signer in your trust store. I would have expected that you fail here if the returned vector does not have length 1 (reason: untrusted cert). If this list has more than one entry, you've got a collision in your hash, and that should never happen, right?

Reply:
A collision in the hash can always happen, it's not that likely but might happen. A `HashedId8` isn't a full SHA-256 hash but only the first 8 bytes (=64 bit) of it. Even with a full hash it might happen theoretically.

You're right that this currently requires the signer to be directly in the trust store. Full chain validation will follow in another PR.

Reply:
The problem is that if it happens, you can't really trust the result anymore either, which is why I suggested length != 1. It has the (very minor) added advantage that the validation time will be more consistent.

Reply:
Yes, it needs some additional API changes to continue the chain checking with the right one in that case.